src/utils/geometry.js

```diff
@@ -28,9 +28,97 @@ export function perpendicular ([x, y]) {
     return [-y, x];
 }
 
+export function getLineNormal (a, b) {
+    const dx = b[0] - a[0];
+    const dy = b[1] - a[1];
+    return normalize([-dy, dx]);
+}
+
+export function getJointNormal (a, b, c) {
+    const u = normalize([a[0] - b[0], a[1] - b[1]]);
+    const v = normalize([c[0] - b[0], c[1] - b[1]]);
+    const sin = -u[1] * v[0] + u[0] * v[1];
+    if (sin !== 0) {
+        return [(u[0] + v[0]) / sin, (u[1] + v[1]) / sin];
+    }
+}
+
+export function normalize (v) {
+    const s = Math.sqrt(v[0] * v[0] + v[1] * v[1]);
+    return [v[0] / s, v[1] / s];
+}
+
+// Returns true if p is inside the triangle or on a triangle's edge, false otherwise
+// Parameters in {x: 0, y:0} form
+export function pointInTriangle (p, v1, v2, v3) {
+    // https://stackoverflow.com/questions/2049582/how-to-determine-if-a-point-is-in-a-2d-triangle
+    // contains an explanation of both this algorithm and one based on barycentric coordinates,
+    // which could be faster, but, nevertheless, it is quite similar in terms of required arithmetic operations
+
+    if (equalPoints(v1, v2) || equalPoints(v2, v3) || equalPoints(v3, v1)) {
+        // Avoid zero area triangle
+        return false;
```

Comment:
return p === v1 ?

Reply:
Could be, but, if that is true, then the point in triangle for another triangle of the feature should return true. I hope there are no features with just a  zero-sized "triangle"